pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/ExactlyOnceMultithreadedProcessor.java

```diff
@@ -0,0 +1,495 @@
+
+                // Create streams.
+                try (StreamManager streamManager = StreamManager.create(controllerURI)) {
+                    streamManager.createScope(scope);
+                    StreamConfiguration streamConfig = StreamConfiguration.builder()
+                            .scalingPolicy(ScalingPolicy.byEventRate(
+                                    Parameters.getTargetRateEventsPerSec(),
+                                    Parameters.getScaleFactor(),
+                                    Parameters.getMinNumSegments()))
+                            .build();
+                    streamManager.createStream(scope, inputStreamName, streamConfig);
+                    // Since we are starting processing from the beginning, delete and create a new output stream.
+                    // TODO: Should we truncate stream instead of deleting?
+                    try {
+                        streamManager.sealStream(scope, outputStreamName);
+                    } catch (Exception e) {
+                        if (!(e.getCause() instanceof InvalidStreamException)) {
+                            throw e;
+                        }
+                    }
+                    // TODO: It would be nice if deleteStream did not require sealStream to be called.
+                    streamManager.deleteStream(scope, outputStreamName);
+                    streamManager.createStream(scope, outputStreamName, streamConfig);
+                }
+
+                // Create a reader group that starts from the earliest event.
+                builder = builder.stream(Stream.of(scope, inputStreamName));
+                startFromCheckpointName = null;
+            }
+
+            clientFactory = ClientFactory.withScope(scope, controllerURI);
+            writer = clientFactory.createEventWriter(
+                    outputStreamName,
+                    new UTF8StringSerializer(),
+                    EventWriterConfig.builder().build());
+
+            // Create a reader group manager. It must remain open to allow manual checkpoints to work.
+            readerGroupManager = ReaderGroupManager.withScope(scope, controllerURI);
+
+            final ReaderGroupConfig readerGroupConfig = builder.build();
+            readerGroupName = UUID.randomUUID().toString().replace("-", "");
+            readerGroupManager.createReaderGroup(readerGroupName, readerGroupConfig);
+            readerGroup = readerGroupManager.getReaderGroup(readerGroupName);
+
+            initiateCheckpointExecutor = Executors.newScheduledThreadPool(1);
+            performCheckpointExecutor = Executors.newScheduledThreadPool(1);
+            workerExecutor = Executors.newFixedThreadPool(numWorkers);
+        }
+
+        /**
+         * Commit all transactions that are part of a checkpoint.
+         *
+         * @param checkpointName
+         */
+        private void commitTransactions(String checkpointName) {
```

Comment:
I do not understand how incremental recovery is possible using the Pravega API. Can you provide a sketch?

Here is my understanding. Please correct me where I am wrong. 
1. With Pravega reader groups, there is no way for an individual reader to obtain it's current offset, as would be possible with Kafka for example. Only the caller of initiateCheckpoint can get the stream cut and it will only get it once all readers have handled the checkpoint.
2. If a transaction is committed before the stream cut can be persisted to storage, then the recovery procedure will not know where to start a new reader group. (Should the recovery procedure use the original reader group instead?)
3. Therefore, we must guarantee the following order: initiateCheckpoint -> workers flush transactions, master waits for all readers by getting the checkpoint future -> master persists stream cut -> commit transactions.
4. Another complication with incremental recovery is that a new reader may not be assigned the same segments as a previously failed reader. This makes it very difficult to recover a consistent snapshot of the state in a stateful processor.

Reply:
Sorry. I find the use of transactions in the discussion confusing. I thought checkpoints and transactions were not necessarily related, but 2 different mechanisms.

Reply:
I find it easiest to think about what is needed to recover from a restart of the processor. The reader group must be positioned at a stream cut. Any transactions flushed prior to the stream cut but not committed must be committed. The transactions are flushed when readers get their checkpoint so that these are exactly aligned.

Reply:
@claudiofahey I think you are making things complicated by focusing on the need to recreate the whole reader group. That is in fact undesirable as it means restarting all the workers because one of them failed. 

Instead let's frame the goal as follows: We're trying to show how to perform exactly once processing. So following a failure reading must resume from the point at which the output was written. 

When a reader fails, readerOffline() can be invoked passing the position the reader failed at and the reader group will redistribute those segments at that position of the remaining readers or any new ones that join the group. 

So if every time a processor commits a transaction it stores the last position object, you can always resume from the correct location. This of course requires you to store the position object. There is a way around this though. Because it is possible that the reader might have persisted nothing before it died it is valid to pass `null` as the position object to `readerOffline()` this will result in it assuming the reader processed up to the last checkpoint but not beyond it. So if the processor calls commit on the transaction when it's reader receives a checkpoint event, and creates a new transaction. Then in response to any failure the correct action would be invoke readerOffline passing null for the position.

Reply:
Ok, I can change the code to reuse the same reader group. That will avoid the need for manual checkpoints.

However, I don't see how to safely avoid the need for a two-phase commit process. If I understand you correctly, you are suggesting:

```
eventRead = reader.readNextEvent(1000);
// Assume eventRead.isCheckpoint().
transaction.commit();
transaction = writer.beginTxn();
// What happens if application crashes here?
// Call readNextEvent to inform reader group that we have processed this checkpoint.
eventRead = reader.readNextEvent(1000);
```

But consider what happens if the application crashes after commit() but before the reader group is updated with readNextEvent.
Upon recovery, the reader group will resume at the previous checkpoint and duplicate records will be created.
To avoid this, we still need to use the two-phase commit process of flush, persist transaction IDs, commit, update reader group.
I may be able to do this with another Pravega stream instead of a shared file system but I don't see how to completely eliminate
this need.